packages/prisma-client/prisma/migrations/20240524155926_patch_map/migration.sql

```diff
@@ -0,0 +1,200 @@
+ * @param {T} draft - The draft object to apply patches to.
+ * @param {ReadonlyArray<Patch>} patches
+ * @returns {T} The patched draft object.
+ */
+const applyPatches = (draft, patches) => {
+  patches.forEach((patch) => {
+    const { path, op } = patch;
+
+    let base = draft;
+    for (let i = 0; i < path.length - 1; i++) {
+      // @ts-ignore
+      const parentType = getArchtype(base);
+      let p = path[i];
+      if (typeof p !== "string" && typeof p !== "number") {
+        p = "" + p;
+      }
+
+      base = get(base, p);
+
+      if (typeof base !== "object") {
+        throw new Error("die");
+      }
+    }
+
+    const type = getArchtype(base);
+    const value = patch.value;
+
+    const key = path[path.length - 1];
+    switch (op) {
+      case "replace":
+        switch (type) {
+          case "Map":
+            // @ts-ignore
+            return base.set(key, value);
+          case "Set":
+            throw new Error("set is not supported");
+          default:
+            // @ts-ignore
+            return (base[key] = value);
+        }
+      case "add":
+        switch (type) {
+          case "Array":
+            // @ts-ignore
+            return key === "-" ? base.push(value) : base.splice(key, 0, value);
+          case "Map":
+            // @ts-ignore
+            return base.set(key, value);
+          case "Set":
+            // @ts-ignore
+            return base.add(value);
+          default:
+            // @ts-ignore
+            return (base[key] = value);
+        }
+      case "remove":
+        switch (type) {
+          case "Array":
+            // @ts-ignore
+            return base.splice(key, 1);
+          case "Map":
+            // @ts-ignore
+            return base.delete(key);
+          case "Set":
+            // @ts-ignore
+            return base.delete(patch.value);
+          default:
+            // @ts-ignore
+            return delete base[key];
+        }
+      default:
+        throw new Error("invalid op");
+    }
+  });
+
+  return draft;
+};
+
+/**
+ *
+ * @param {string} dataStr
+ * @param {string} primaryKeyCommaSeparated - if empty then it's js array or plain object, otherwise is a map
+ * @param {string} patchesStr
+ * @returns {string}
+ */
+const patch = (dataStr, primaryKeyCommaSeparated, patchesStr) => {
+  /**
+   * @param {*} item
+   * @returns
+   */
+
+  const primaryKeyArray = primaryKeyCommaSeparated
+    .split(",")
+    .map((item) => item.trim());
+
+  const isPlainObjectOrArray =
+    primaryKeyArray.length === 1 && primaryKeyArray[0] === "";
+
+  /**
+   * @type { undefined | ((item: any) => string) }
+   */
+  let getKey = isPlainObjectOrArray
+    ? undefined
+    : (item) => {
+        if (primaryKeyArray.length === 1) {
+          return item[primaryKeyArray[0]];
+        }
+
+        return primaryKeyArray
+          .map((key) => (item[key] == null ? "" : item[key]))
+          .join(":");
+      };
+
+  let data = JSON.parse(dataStr);
```

Comment:
shoudln't this be in try/catch?

Reply:
why? let it fail

Reply:
do we have ways to handle that case?

Reply:
What the reason? Its perfectly ok if this function will fail on json parse. Without any wrapper.

Reply:
so what happens it it fails, did you try? where does error land? are you catching it somewhere in prisma then as a usual postgres error? or is it blowing up and doing some weird shit)

Reply:
This doesn't change any behavioir with SQL statements we did before. What if any previous SQL fail. It's not the refactor of how errors from SQL are shown on UI or console.  

Will be shown this
```
"errors": "\nInvalid `prisma.$executeRaw()` invocation:\n\n\nRaw query failed. Code: `XX000`. Message: `ERROR: SyntaxError: Unexpected number in JSON at position 1`"
```
As like as if any SQL statement before would fail.

And it's ok as there is 0 chance to get error like this without data corruption on db level what is not we are expecting when writing the code.